src/System.Diagnostics.Process/tests/ProcessStartInfoTests.cs

```diff
@@ -185,70 +185,52 @@ public void TestEnvironmentProperty()
+            Environment.SetEnvironmentVariable(ExtraEnvVar, "\x1234" + Environment.NewLine + "\x5678"); // ensure some Unicode characters and newlines are in the output
             try
             {
-                var expectedEnv = new HashSet<string>();
-                var actualEnv = new HashSet<string>();
+                // Schedule a process to see what env vars it gets.  Have it write out those variables
+                // to its output stream so we can read them.
                 Process p = CreateProcess(() =>
                 {
-                    foreach (DictionaryEntry envVar in Environment.GetEnvironmentVariables())
-                    {
-                        Console.WriteLine(envVar.Key + "=" + envVar.Value);
-                    }
-
+                    Console.Write(string.Join(ItemSeparator, Environment.GetEnvironmentVariables().Cast<DictionaryEntry>().Select(e => e.Key + "=" + e.Value)));
                     return SuccessExitCode;
                 });
                 p.StartInfo.StandardOutputEncoding = Encoding.UTF8;
-
                 p.StartInfo.RedirectStandardOutput = true;
-                p.OutputDataReceived += (s, e) =>
-                {
-                    if (e.Data != null)
-                    {
-                        expectedEnv.Add(e.Data);
-                    }
-                };
-
                 p.Start();
-                p.BeginOutputReadLine();
-
-                foreach (KeyValuePair<string, string> envVar in p.StartInfo.Environment)
-                {
-                    actualEnv.Add(envVar.Key + "=" + envVar.Value);
-                }
-
+                string output = p.StandardOutput.ReadToEnd();
                 Assert.True(p.WaitForExit(WaitInMS));
-                p.WaitForExit(); // This ensures async event handlers are finished processing.
 
-                // Validate against StartInfo.Environment
-                if (!expectedEnv.SetEquals(actualEnv))
-                {
-                    var expected = string.Join(", ", expectedEnv.Except(actualEnv));
-                    var actual = string.Join(", ", actualEnv.Except(expectedEnv));
+                // Parse the env vars from the child process
+                var actualEnv = new HashSet<string>(output.Split(new[] { ItemSeparator }, StringSplitOptions.None));
 
-                    Assert.True(false, string.Format("Expected: {0}{1}Actual: {2}", expected, Environment.NewLine, actual));
+                // Validate against StartInfo.Environment.
+                var startInfoEnv = new HashSet<string>(p.StartInfo.Environment.Select(e => e.Key + "=" + e.Value));
+                if (!startInfoEnv.SetEquals(actualEnv))
+                {
+                    Assert.True(false, string.Format("Expected: {0}{1}Actual: {2}",
+                        string.Join(", ", startInfoEnv.Except(actualEnv)), 
+                        Environment.NewLine,
+                        string.Join(", ", actualEnv.Except(startInfoEnv))));
```

Comment:
Ah yeah, I missed that this just logs the condition. We could still get rid of the string.Format by doing Assert.NotEqual on the strings, but I don't care much.